### PR TITLE
Add output bundle spec to common module

### PR DESCRIPTION
**Pull request by Yuangwang**
This interface will be used in the other adapters as the output bundle

### CODE REVIEW

File: packages/@apphosting/common/src/index.ts [hunk 1]
```diff
@@ -1,5 +1,58 @@
...
+// Additonal fields needed for adapter identification
+interface Metadata {
+  // Name of the adapter (this should be the npm package name)
+  adapterNpmPackageName: string;
+  // Version of the adapter
+  adapterVersion?: string;
+  // Name of the framework that is being supported
+  framework: string;
+  // Version of the framework that is being supported
+  frameworkVersion?: string;
```
[the remark this answers is not in the document]
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts:43]: Yes that's exactly right

File: packages/@apphosting/common/src/index.ts [hunk 2]
```diff
@@ -1,5 +1,58 @@
 import { spawn } from "child_process";
 
+// Output bundle specifications to be written to bundle.yaml
+export interface OutputBundle {
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:4]: So one quick thought: I thought OutputBundle was the name for the entirety of the framework build process which includes:
* bundle.yaml
* static assets
* source build artifacts

Is it accurate to name this outputbundle when we're actually defining the metadata for the output bundle?

File: packages/@apphosting/common/src/index.ts [hunk 3]
```diff
@@ -1,5 +1,58 @@
...
+  version: "v1alpha";
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:5]: prefer to not use alpha/beta versioning. it's being phased out at google and it's added overhead to our own internal api release processes

File: packages/@apphosting/common/src/index.ts [hunk 4]
```diff
@@ -1,5 +1,58 @@
...
+  serverConfig: ServerConfig;
+  metadata: Metadata;
+}
+
+// Fields needed to configure the App Hosting server
+interface ServerConfig {
+  // Command to start the server (e.g. ["node", "dist/index.js"]) assume this command is run from the root dir of the workspace
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:12]: punctation nit: "(eg ...). Assume"

File: packages/@apphosting/common/src/index.ts [hunk 5]
```diff
@@ -1,5 +1,58 @@
...
+  runCommand: string[];
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:13]: ooc why string[] and not string?

File: packages/@apphosting/common/src/index.ts [hunk 6]
```diff
@@ -1,5 +1,58 @@
...
+  // Environment variable present in the server execution environment.
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:14]: Is this implying Cloud Run (eg runtime)? Do we need to change this comment if we ever want to support buildtime?

File: packages/@apphosting/common/src/index.ts [hunk 7]
```diff
@@ -1,5 +1,58 @@
...
+  environmentVariables?: EnvVarConfig[];
+  // The maximum number of concurrent requests that each server instance can receive.
+  // Defaults to 80.
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:17]: Do we want to drop these defaults now that we know GCP can adjust default values per user?

File: packages/@apphosting/common/src/index.ts [hunk 8]
```diff
@@ -1,5 +1,58 @@
...
+  concurrency?: number;
+  // The number of CPUs used in a single server instance.
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:19]: Are we intentionally diverging from our server docs?  https://cloud.google.com/run/docs/reference/rest/v2/projects.locations.services#resource:-service

File: packages/@apphosting/common/src/index.ts [hunk 9]
```diff
@@ -1,5 +1,58 @@
...
+  memoryMiB?: string;
+  // The limit on the minimum number of function instances that may coexist at a given time.
+  // Defaults to 0.
+  minInstances?: number;
+  // The limit on the maximum number of function instances that may coexist at a given time.
+  // Defaults to 100.
+  maxInstances?: number;
+}
+
+// Additonal fields needed for adapter identification
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:34]: "adapter identification" doesn't feel comprehensive here

File: packages/@apphosting/common/src/index.ts [hunk 10]
```diff
@@ -1,5 +1,58 @@
...
+interface Metadata {
+  // Name of the adapter (this should be the npm package name)
+  adapterNpmPackageName: string;
+  // Version of the adapter
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:38]: Suggestion to include example strings in these docs, eg: 

"Name of the framework that is being supported, e.g. "angular""
"Version of the framework that is being supported, e.g. "18.0.1""

File: packages/@apphosting/common/src/index.ts [hunk 11]
```diff
@@ -1,5 +1,58 @@
...
+}
+
+// Configs necessary to define environment variables
+interface EnvVarConfig {
+  // Name of the variable
+  variable: string;
+  // Value associated with the variable
+  value: string;
+  // Where the variable will be available, for now this will always be RUNTIME
+  availability: "RUNTIME"[];
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:53]: string or enum?

File: packages/@apphosting/common/src/index.ts [hunk 12]
```diff
@@ -1,5 +1,58 @@
...
+  adapterNpmPackageName: string;
+  // Version of the adapter
+  adapterVersion?: string;
+  // Name of the framework that is being supported
+  framework: string;
+  // Version of the framework that is being supported
+  frameworkVersion?: string;
+}
+
+// Configs necessary to define environment variables
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:46]: Consider: "Represents a single environment variable."

File: packages/@apphosting/common/src/index.ts (hunk 5 above)
Comment (tonyjhuang) [reply packages/@apphosting/common/src/index.ts:13]: Hmm okay I personally would recommend we prioritize whatever makes for a better DX for adapter authors here. Splitting by " " isn't a ton of overhead on the buildpacks side, but adapter authors will have to pay the cost of consuming this API every time they add a new framework

File: packages/@apphosting/common/src/index.ts (hunk 7 above)
Comment (tonyjhuang) [reply packages/@apphosting/common/src/index.ts:17]: What if we link to the product docs instead so we have a single source of truth?

File: packages/@apphosting/common/src/index.ts [hunk 13]
```diff
@@ -31,26 +31,31 @@ interface ServerConfig {
...
-// Additonal fields needed for adapter identification
+// Additonal fields needed for identifying the framework and adapter being used
 interface Metadata {
-  // Name of the adapter (this should be the npm package name)
-  adapterNpmPackageName: string;
-  // Version of the adapter
+  // Name of the adapter (this should be the official package name) e.g. "@apphosting/adapter-nextjs"
+  adapterPackageName: string;
+  // Version of the adapter, e.g. "18.0.1"
   adapterVersion?: string;
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:39]: why optional?

File: packages/@apphosting/common/src/index.ts [hunk 14]
```diff
@@ -31,26 +31,31 @@ interface ServerConfig {
...
   // Value associated with the variable
   value: string;
-  // Where the variable will be available, for now this will always be RUNTIME
-  availability: "RUNTIME"[];
+  // Where the variable will be available, for now only RUNTIME is supported
+  availability: Availabilities.Runtime[];
+}
+
+// Represents where environment variables are available from
+enum Availabilities {
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:57]: enum Availability

File: packages/@apphosting/common/src/index.ts [hunk 15]
```diff
@@ -31,26 +31,31 @@ interface ServerConfig {
...
   variable: string;
   // Value associated with the variable
   value: string;
-  // Where the variable will be available, for now this will always be RUNTIME
-  availability: "RUNTIME"[];
+  // Where the variable will be available, for now only RUNTIME is supported
+  availability: Availabilities.Runtime[];
+}
+
+// Represents where environment variables are available from
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:56]: comment nit: this is a bit confusing, how about "Represents where environment variables are made available."

File: packages/@apphosting/common/src/index.ts [hunk 16]
```diff
@@ -31,26 +31,31 @@ interface ServerConfig {
...
+enum Availabilities {
+  Runtime,
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:58]: We should include a doc string here

File: packages/@apphosting/common/src/index.ts [hunk 17]
```diff
@@ -1,17 +1,17 @@
 import { spawn } from "child_process";
 
-// Output bundle specifications to be written to bundle.yaml
-export interface OutputBundle {
-  version: "v1alpha";
+// Output bundle metadata specifications to be written to bundle.yaml
+export interface OutputBundleConfigs {
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:4]: nit: OutputBundleConfig

**Review by tonyjhuang — CHANGES_REQUESTED** [review]

File: packages/@apphosting/common/src/index.ts (hunk 5 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts]: Sure that makes sense, changed this to string

File: packages/@apphosting/common/src/index.ts (hunk 7 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts]: Removed the defaults in favor of linking the docs.

File: packages/@apphosting/common/src/index.ts (hunk 8 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts:19]: Oh good call, yeah this shouldn't diverge from those. Changed CPU to number

File: packages/@apphosting/common/src/index.ts (hunk 13 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts]: Originally this field was added so authors had the option to tell us what version if they wanted to (technically this is something we can self discover on the buildpacks side too) but I think it's pretty straightforward for adapter authors to provide this to us as well (done here as an example https://github.com/FirebaseExtended/firebase-framework-tools/pull/248/files#diff-c4645e334899a380f959aecbba2ce11a7ffcd895b12daed80af12d265eb9d437R120-R125) 

I've gone ahead and made it required so we don't need to do it in the buildpacks anymore

File: packages/@apphosting/common/src/index.ts (hunk 14 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts]: Done

File: packages/@apphosting/common/src/index.ts (hunk 16 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts:54]: Done

File: packages/@apphosting/common/src/index.ts (hunk 15 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts]: Done

File: packages/@apphosting/common/src/index.ts (hunk 17 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts]: Done

File: packages/@apphosting/common/src/index.ts [hunk 18]
```diff
@@ -1,56 +1,57 @@
...
-  // Defaults to 80.
   concurrency?: number;
   // The number of CPUs used in a single server instance.
-  // Defaults to 1.
-  cpu?: "fractional" | 1 | 2 | 4 | 6 | 8;
+  cpu?: number;
   // The amount of memory available for a server instance.
-  // Commonly one of: "256" | "512" | "1024" | "2048" | "4096" | "8192" | "16384"
-  // Defaults to 512MiB.
   memoryMiB?: string;
```
Comment (tonyjhuang) [inline packages/@apphosting/common/src/index.ts:22]: why string?

**Review by tonyjhuang — APPROVED** [review]

File: packages/@apphosting/common/src/index.ts (hunk 18 above)
Comment (Yuangwang) [reply packages/@apphosting/common/src/index.ts:22]: Good catch, we probably want number here too

**Pull request merged**